Generate error on unknown critical extension during path validation

cordney opened the pull request:
Previously unknown critical extensions were rejected during X509_Certificate constructor, which inhibited inspecting other parts of such a certificate. Refactored the certificate extensions code so that the path validation routine performs this check only. Additionally, added an interface for extensions to inspect the path during path validation.

* Moved some includes and forward declarations to avoid a circular dependency of `X509_Certificate` and  `Certificate_Extension`
* Change `X509_Certificate::v3_extensions()` to return an `Extensions` object (needed to add explicit copy constructor and assignment operator)
* Add new optional `validate()` on `Certificate_Extension` interface for extension to inspect the path during path validation
* Move name constraints checking code from `check_chain()` to `Name_Constraints::validate()`
* Added `Unknown_Critical_Extension` class that always adds a failure to the path validation result

Fixes GH #449.

cordney commented:
One thing to add: To test, there is a x509test test case with an unknown critical extension in it. This is not part of the x509test suite in botan, yet. We would have added it, but the root CA key was not in the botan repo, so we could not create it.

cordney commented:
Added the `Unknown_Critical_Extension` class, which always adds a failure to the path validation result.
Also marked this PR as WIP. I wanted to hear your opinion first.

randombit commented on src/lib/cert/x509/x509_ext.cpp:
```diff
@@ -40,7 +40,7 @@ Certificate_Extension* Extensions::get_extension(const OID& oid)
    X509_EXTENSION("X509v3.ReasonCode", CRL_ReasonCode);
 
-   return nullptr;
+   return critical ? new Cert_Extension::Unknown_Critical_Extension() : nullptr;
```
Capturing the OID might be useful for diagnostics.

cordney replied:
Added, thx.

randombit commented:
Don't have time for a full review right now, but yes this matches what I had in mind.

cordney commented on src/lib/cert/x509/x509_ext.cpp:
```diff
@@ -136,7 +146,7 @@ void Extensions::decode_from(BER_Decoder& from_source)
-      std::unique_ptr<Certificate_Extension> ext(get_extension(oid));
+      std::unique_ptr<Certificate_Extension> ext(get_extension(oid, critical));
 
       if(!ext && critical && m_throw_on_unknown_critical)
```
Because of `get_extension()` now returning an `Unknown_Critical_Extension`, this check does not work anymore. I wonder if this functionality is neither desired nor needed anymore?

randombit replied:
At first glance I would agree, this check is now needless as validation will fail later on. However this does remove all uses of `m_throw_on_unknown_critical` from `Extensions`. This is used by CRLs and in fact I see the default parameter in `X509_CRL`s constructor has this set to `false` which seems wrong to me. Processing the CRL in the face of missing information (which a critical extension surely represents, else it wouldn't be marked critical) is probably a bad idea. So with this change, all such `m_throw_on_unknown_critical` flags in x509 (crl_ent.cpp, x509_crl.cpp) can be removed.

cordney commented:
I moved the name constraints checking code by @flanfly into the extension and cleaned up the code. I also updated the description to list all changes made in this PR. Holding the DN directly in the certificate is not in this PR now, I think we should add it to the TODOs.

cordney commented:
Thanks for merging. Sorry, I did not have time to finish this until the weekend. Will you make the remaining changes or are you expecting another PR?

randombit commented:
I was planning on removing the now vestigial `m_throw_on_unknown_critical` flags and moving the other validation checks (basic constraints, etc) from path processing into the extension checks. Also there needs to be a test to confirm unknown critical extensions are handled as intended; I found https://bugzilla.mozilla.org/show_bug.cgi?id=1009161 which has a good example. Past these items is there other work related to this PR still to be done?

cordney commented:
I already gave putting the basic constraints extension checks into the `Extension` a try, but failed. This is a really special extension in that it depends on the key usage bits and there are other special checks that RFC 5280 requires, like `reject v1 and v2 CA certs` where the basic constraints extension is not present. I ended up having some checks in the extension, but still others in `check_chain()`, so I dropped this commit. But maybe you come up with a better implementation.

The pkix issue is a good example. And part of the problem of what is behind #449. We issued certs with this extension in one of our PKIs until two years ago and are now moving to Botan. Until now, I managed refuse implementing this 90ies extension in Botan and so did the Mozilla devs in pkix initially. But in the end they reimplement it anyway, because it caused too many problems with legacy certificates on the Internet PKI.
